sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala

```diff
@@ -249,10 +247,20 @@ abstract class HashExpression[E] extends Expression {
 
   override def nullable: Boolean = false
 
+  private def hasMapType(dt: DataType): Boolean = {
+    dt.existsRecursively(_.isInstanceOf[MapType])
+  }
+
   override def checkInputDataTypes(): TypeCheckResult = {
     if (children.length < 1) {
       TypeCheckResult.TypeCheckFailure(
         s"input to function $prettyName requires at least one argument")
+    } else if (children.forall(child => hasMapType(child.dataType)) &&
+      !SQLConf.get.getConf(SQLConf.LEGACY_USE_HASH_ON_MAPTYPE)) {
+      TypeCheckResult.TypeCheckFailure(
+        s"input to function $prettyName cannot contain elements of MapType. Logically same maps " +
+          "may have different hashcode, thus hash expressions are prohibited on MapType " +
+          "elements. To restore previous behavior set spark.sql.legacy.useHashOnMapType to true.")
```

Comment:
nit: `spark.sql.legacy.useHashOnMapType` -> `SQLConf.LEGACY_USE_HASH_ON_MAPTYPE.key`

Reply:
I used `spark.sql.legacy.useHashOnMapType` to maintain consistency with what i updated in migration guide.

Reply:
We should not hardcode the config name. Better to use `${SQLConf.LEGACY_USE_HASH_ON_MAPTYPE.key}`